### PR TITLE
Support configuring ScyllaDB and Scylla Manager versions for e2e tests with flags

### LINKED ISSUE #2320: Support configuring ScyllaDB and Scylla Manager versions for e2e tests with flags
We need to set up early testing against ScyllaDB and Manager pre-release so we need to have this configurable when running `scylla-operator-tests`. These flags should override the default version defined in the repository.

Ref https://github.com/scylladb/scylla-operator/issues/1937

### PR DESCRIPTION
**Pull request by grzywin**
**Description of your changes:**
This PR makes the flags for ScyllaDB and Scylla Manager versions configurable. Hardcoded values (versions) in config.yaml will be overwritten by the ones passed as arguments.

**Which issue is resolved by this Pull Request:**
Resolves #2320

### CODE REVIEW

File: pkg/cmd/tests/options.go [hunk 1]
```diff
@@ -164,6 +181,28 @@ func (o *TestFrameworkOptions) Validate(args []string) error {
 		errors = append(errors, fmt.Errorf("gcs-service-account-key-path and s3-credentials-file-path can't be set simultanously"))
 	}
 
+	versionPattern := regexp.MustCompile(`^\d+\.\d+\.\d+(-[a-zA-Z0-9]+)?$`)
```
Comment (rzetelskik) [inline pkg/cmd/tests/options.go:184]: This doesn't cover tag + digest entirely (e.g. latest wouldn't be caught by that), see the following for reference:
- https://github.com/distribution/reference/blob/8c942b0459dfdcc5b6685581dd0a5a470f615bff/reference.go#L6C44-L6C70
- https://github.com/distribution/reference/blob/8c942b0459dfdcc5b6685581dd0a5a470f615bff/regexp.go#L39
- https://github.com/distribution/reference/blob/8c942b0459dfdcc5b6685581dd0a5a470f615bff/regexp.go#L9
I'd import `distribution/reference` to handle that and look there to see how to match and capture both correctly. It might differ between ScyllaDB and manager versions. From the top of my head I'm not sure we handle [tag + ":"] + digest for scylla versions - you'll have to try.

File: hack/.ci/lib/e2e.sh [hunk 2]
```diff
@@ -274,6 +274,11 @@ spec:
     - "--object-storage-bucket=${SO_BUCKET_NAME}"
     - "--gcs-service-account-key-path=${gcs_sa_in_container_path}"
     - "--s3-credentials-file-path=${s3_credentials_in_container_path}"
+    - "--scylla-db-version=${SCYLLA_DB_VERSION}"
+    - "--scylla-manager-version=${SCYLLA_MANAGER_VERSION}"
+    - "--scylla-manager-agent-version=${SCYLLA_MANAGER_AGENT_VERSION}"
+    - "--scylla-update-from=${SCYLLA_UPDATE_FROM}"
+    - "--scylla-upgrade-from=${SCYLLA_UPGRADE_FROM}"
```
Comment (rzetelskik) [inline hack/.ci/lib/e2e.sh:281]: You have to declare these env vars, otherwise the script will fail on undeclared

File: hack/ci-deploy.sh [hunk 3]
```diff
@@ -113,6 +113,19 @@ if [[ -n "${SO_SCYLLACLUSTER_STORAGECLASS_NAME:-}" ]]; then
 elif [[ -n "${SO_SCYLLACLUSTER_STORAGECLASS_NAME+x}" ]]; then
   yq e --inplace 'del(.spec.datacenter.racks[0].storage.storageClassName)' "${DEPLOY_DIR}/manager/50_scyllacluster.yaml"
 fi
+
+if [[ -n "${SCYLLA_DB_VERSION:-}" ]]; then
```
Comment (rzetelskik) [inline hack/ci-deploy.sh:117]: SCYLLADB_VERSION

File: pkg/cmd/tests/options.go [hunk 4]
```diff
@@ -118,6 +130,11 @@ func (o *TestFrameworkOptions) AddFlags(cmd *cobra.Command) {
 	cmd.PersistentFlags().StringVarP(&o.ObjectStorageBucket, "object-storage-bucket", "", o.ObjectStorageBucket, "Name of the object storage bucket.")
 	cmd.PersistentFlags().StringVarP(&o.GCSServiceAccountKeyPath, "gcs-service-account-key-path", "", o.GCSServiceAccountKeyPath, "Path to a file containing a GCS service account key.")
 	cmd.PersistentFlags().StringVarP(&o.S3CredentialsFilePath, "s3-credentials-file-path", "", o.S3CredentialsFilePath, "Path to the AWS credentials file providing access to the S3 bucket.")
+	cmd.PersistentFlags().StringVarP(&o.ScyllaDBVersion, "scylla-db-version", "", o.ScyllaDBVersion, "Version of ScyllaDB to use in tests.")
```
Comment (rzetelskik) [inline pkg/cmd/tests/options.go:133]: s/scylla-db-version/scylladb-version

File: pkg/cmd/tests/options.go [hunk 5]
```diff
@@ -118,6 +130,11 @@ func (o *TestFrameworkOptions) AddFlags(cmd *cobra.Command) {
...
+	cmd.PersistentFlags().StringVarP(&o.ScyllaManagerVersion, "scylla-manager-version", "", o.ScyllaManagerVersion, "Version of Scylla Manager to use in tests.")
+	cmd.PersistentFlags().StringVarP(&o.ScyllaManagerAgentVersion, "scylla-manager-agent-version", "", o.ScyllaManagerAgentVersion, "Version of Scylla Manager Agent to use in tests.")
+	cmd.PersistentFlags().StringVarP(&o.ScyllaDBUpdateFrom, "scylla-update-from", "", o.ScyllaDBUpdateFrom, "Version of Scylla Update to use in tests.")
```
Comment (rzetelskik) [inline pkg/cmd/tests/options.go:136]: s/scylla-update-from/scylladb-update-from-version

File: pkg/cmd/tests/options.go [hunk 6]
```diff
@@ -118,6 +130,11 @@ func (o *TestFrameworkOptions) AddFlags(cmd *cobra.Command) {
...
+	cmd.PersistentFlags().StringVarP(&o.ScyllaDBUpgradeFrom, "scylla-upgrade-from", "", o.ScyllaDBUpgradeFrom, "Version of Scylla Upgrade to use in tests.")
```
Comment (rzetelskik) [inline pkg/cmd/tests/options.go:137]: s/scylla-upgrade-from/scylladb-upgrade-from-version

File: pkg/cmd/tests/options.go (hunk 5 above)
Comment (rzetelskik) [reply pkg/cmd/tests/options.go:136]: also `Version of Scylla Update to use in tests.` doesn't make much sense to me (applies to upgrade as well), how about "Version of ScyllaDB to update from"?

File: pkg/cmd/tests/options.go (hunk 4 above)
Comment (rzetelskik) [reply pkg/cmd/tests/options.go:133]: nit: You don't need to say "in tests" either, it's pretty clear as it's a test runner command (applies to all flags)

File: pkg/cmd/tests/options.go (hunk 5 above)
Comment (rzetelskik) [reply pkg/cmd/tests/options.go]: still applies partially, s/scylla/scylladb

File: pkg/cmd/tests/options.go [hunk 7]
```diff
@@ -164,6 +181,33 @@ func (o *TestFrameworkOptions) Validate(args []string) error {
 		errors = append(errors, fmt.Errorf("gcs-service-account-key-path and s3-credentials-file-path can't be set simultanously"))
 	}
 
+	if o.ScyllaDBVersion != "" {
```
Comment (rzetelskik) [inline pkg/cmd/tests/options.go:184]: why only do this conditionally? Empty string is an invalid "version", no?

File: pkg/cmd/tests/options.go [hunk 8]
```diff
@@ -164,6 +181,33 @@ func (o *TestFrameworkOptions) Validate(args []string) error {
...
+		if _, err := reference.ParseAnyReference(o.ScyllaDBVersion); err != nil {
```
Comment (rzetelskik) [inline pkg/cmd/tests/options.go:185]: this is incorrect, ParseAnyReference is going to "accept" (as in not return an error) much more than just `tag ['@' digest]`. That way you'd accept a whole reference `docker.io/scylladb/scylladb:latest`, just a name `docker.io/scylladb/scylladb`, just a remote name `scylladb/scylladb`, just a path component `scylladb` etc. 

We want to validate this matches `tag ['@' digest]`.

Again, see https://github.com/scylladb/scylla-operator/pull/2355#discussion_r1935998932 with references.

File: pkg/cmd/tests/options.go [hunk 9]
```diff
@@ -186,6 +230,22 @@ func (o *TestFrameworkOptions) Complete(args []string) error {
...
+	}
+	if o.ScyllaManagerAgentVersion == "" {
+		o.ScyllaManagerAgentVersion = configassets.Project.Operator.ScyllaDBManagerAgentVersion
+	}
+	if o.ScyllaDBUpdateFrom == "" {
+		o.ScyllaDBUpdateFrom = configassets.Project.OperatorTests.ScyllaDBVersions.UpdateFrom
+	}
+	if o.ScyllaDBUpgradeFrom == "" {
+		o.ScyllaDBUpgradeFrom = configassets.Project.OperatorTests.ScyllaDBVersions.UpgradeFrom
+	}
```
Comment (rzetelskik) [inline pkg/cmd/tests/options.go:247]: you already assign defaults in the constructor, this is redundant

**Review by rzetelskik — COMMENTED** [review]
@grzywin thanks for the updates.

Please don't resolve comments - it should be done by the reviewer to track what was actually done.

Also the commit name "Update generated" is misleading - you're updating dependencies, not generated files.

File: hack/.ci/lib/e2e.sh [hunk 10]
```diff
@@ -234,6 +234,12 @@ function run-e2e {
   ingress_custom_annotations='haproxy.org/ssl-passthrough=true,route.openshift.io/termination=passthrough'
   ingress_controller_address="$( kubectl -n=haproxy-ingress get svc haproxy-ingress --template='{{ .spec.clusterIP }}' ):9142"
 
+  SCYLLADB_VERSION="${SCYLLADB_VERSION:-}"
+  SCYLLA_MANAGER_VERSION="${SCYLLA_MANAGER_VERSION:-}"
+  SCYLLA_MANAGER_AGENT_VERSION="${SCYLLA_MANAGER_AGENT_VERSION:-}"
+  SCYLLA_UPDATE_FROM_VERSION="${SCYLLA_UPDATE_FROM_VERSION:-}"
```
Comment (rzetelskik) [inline hack/.ci/lib/e2e.sh:240]: s/SCYLLA/SCYLLADB

File: hack/.ci/lib/e2e.sh [hunk 11]
```diff
@@ -234,6 +234,12 @@ function run-e2e {
...
+  SCYLLA_UPGRADE_FROM_VERSION="${SCYLLA_UPGRADE_FROM_VERSION:-}"
```
Comment (rzetelskik) [inline hack/.ci/lib/e2e.sh:241]: s/SCYLLA/SCYLLADB

File: pkg/cmd/tests/options.go (hunk 9 above)
Comment (grzywin) [reply pkg/cmd/tests/options.go:247]: @rzetelskik Tests on CI are failing without this part of code. I also ran some tests locally and without it, when we pass empty value to parameter e.g. `scylladb-version=` (like it is in CI) the tests are failing.

File: pkg/cmd/tests/options.go (hunk 7 above)
Comment (grzywin) [reply pkg/cmd/tests/options.go:184]: On CI the test are triggered in following form:
`I0131 22:48:32.144278       1 tests/tests_run.go:358] "Started process" ID=1 Command="/usr/bin/scylla-operator-tests run scylla-operator/conformance/multi-datacenter-parallel --kubeconfig=/var/run/secrets/kubeconfigs/europe-west4.kubeconfig,/var/run/secrets/kubeconfigs/europe-west3.kubeconfig,/var/run/secrets/kubeconfigs/europe-west1.kubeconfig --loglevel=2 --color=false --artifacts-dir=/tmp/artifacts --parallelism=0 --feature-gates=AllAlpha=true,AllBeta=true --ingress-controller-address=10.107.206.204:9142 --ingress-controller-ingress-class-name=haproxy --ingress-controller-custom-annotations=haproxy.org/ssl-passthrough=true,route.openshift.io/termination=passthrough --scyllacluster-node-service-type=Headless --scyllacluster-nodes-broadcast-address-type=PodIP --scyllacluster-clients-broadcast-address-type=PodIP --scyllacluster-storageclass-name=scylladb-local-xfs --object-storage-bucket= --gcs-service-account-key-path= --s3-credentials-file-path= --scylladb-version= --scylla-manager-version= --scylla-manager-agent-version= --scylla-update-from-version= --scylla-upgrade-from-version= --parallel-server-address=127.0.0.1:40981 --loglevel=0 --random-seed=1738363712 --parallel-shard=1"`

The parameters have an empty value "--scylladb-version= " so without this "if not empty" condition assertion will fail and we will get `invalid scylladb-version format`

File: pkg/cmd/tests/options.go (hunk 8 above)
Comment (grzywin) [reply pkg/cmd/tests/options.go:185]: I will take a closer look into this.

Comment (rzetelskik) [reply pkg/cmd/tests/options.go:185]: I see `distribution/reference` doesn't export the actual expressions for tag and digest but only the compiled structs, so you can either import the compiled vars and get the expression with `.String()`
```
regexp.MustCompile("^" + reference.TagRegexp.String() + "(?:" + reference.DigestRegexp.String() + ")?$")
```

or, simpler, just provide references in comments:

```
const (
  // https://github.com/distribution/reference/blob/8c942b0459dfdcc5b6685581dd0a5a470f615bff/regexp.go#L68.
  referenceTagRegexp = `[\w][\w.-]{0,127}`
  
  ...
)

var tagWithOptionalDigestRegexp = regexp.MustCompile("^" + referenceTagRegexp + "(?:" + referenceDigestRegexp + ")?$")
```

Either works, second one avoids the back and forth so it's maybe a bit better.
Verify the regex string though, I wrote it from the top of my head and I'm not sure it even compiles.

File: pkg/cmd/tests/options.go (hunk 7 above)
Comment (rzetelskik) [reply pkg/cmd/tests/options.go:184]: > On CI the test are triggered in following form:

That's how you set it up in this PR, it didn't exist before :wink: 
Empty string **is** an invalid version - we expect `tag ['@' digest]`. We provide a default for an unset flag, but if it's explicitly set to an empty value, then we shouldn't override it but just fail.

So what we need to do instead is to only add the flags conditionally in the e2e runner or set the env var to the defaults coming from the config file. Either way this should happen in the scripts, not in the test runner.

File: pkg/cmd/tests/options.go (hunk 9 above)
Comment (rzetelskik) [reply pkg/cmd/tests/options.go:247]: See https://github.com/scylladb/scylla-operator/pull/2355#discussion_r1939313680

File: hack/.ci/lib/e2e.sh (hunk 11 above)
Comment (grzywin) [reply hack/.ci/lib/e2e.sh]: fixed

File: hack/.ci/lib/e2e.sh (hunk 10 above)
Comment (grzywin) [reply hack/.ci/lib/e2e.sh]: fixed

File: pkg/cmd/tests/options.go (hunk 9 above)
Comment (grzywin) [reply pkg/cmd/tests/options.go]: removed

File: pkg/cmd/tests/options.go (hunk 8 above)
Comment (grzywin) [reply pkg/cmd/tests/options.go]: I removed https://github.com/distribution/reference dependency and used regex directly in our code as in option 2 which you showed. I checked multiple formats:
image

File: pkg/cmd/tests/options.go (hunk 7 above)
Comment (grzywin) [reply pkg/cmd/tests/options.go]: Now, when you are writing about it sound obvious ;) Fixed (I edited e2e.sh)

File: pkg/cmd/tests/options.go [hunk 12]
```diff
@@ -164,6 +181,31 @@ func (o *TestFrameworkOptions) Validate(args []string) error {
 		errors = append(errors, fmt.Errorf("gcs-service-account-key-path and s3-credentials-file-path can't be set simultanously"))
 	}
 
+	const referenceTagRegexp = `[\w][\w.-]{0,127}`
+	const digestRegexp = `[A-Za-z][A-Za-z0-9]*(?:[-_+.][A-Za-z][A-Za-z0-9]*)*[:][[:xdigit:]]{32,}`
+
```
Comment (rzetelskik) [inline pkg/cmd/tests/options.go:186]: this needs a reference to the source (as a comment, I put it in the code snippet earlier for a reason)

Comment (rzetelskik) [reply pkg/cmd/tests/options.go:186]: nit: could be moved to a single const block at the top of the file

File: hack/.ci/lib/e2e.sh [hunk 13]
```diff
@@ -234,6 +234,14 @@ function run-e2e {
...
   ingress_controller_address="$( kubectl -n=haproxy-ingress get svc haproxy-ingress --template='{{ .spec.clusterIP }}' ):9142"
 
+  CONFIG_FILE="$(realpath "$(dirname "${BASH_SOURCE[0]}")/../../../assets/config/config.yaml")"
+
+  SCYLLADB_VERSION="${SCYLLADB_VERSION:-$(yq '.operator.scyllaDBVersion' "$CONFIG_FILE")}"
+  SCYLLADB_MANAGER_VERSION="${SCYLLADB_MANAGER_VERSION:-$(yq '.operator.scyllaDBManagerVersion' "$CONFIG_FILE")}"
+  SCYLLADB_MANAGER_AGENT_VERSION="${SCYLLADB_MANAGER_AGENT_VERSION:-$(yq '.operator.scyllaDBManagerAgentVersion' "$CONFIG_FILE")}"
+  SCYLLADB_UPDATE_FROM_VERSION="${SCYLLADB_UPDATE_FROM_VERSION:-$(yq '.operatorTests.scyllaDBVersions.updateFrom' "$CONFIG_FILE")}"
+  SCYLLADB_UPGRADE_FROM_VERSION="${SCYLLADB_UPGRADE_FROM_VERSION:-$(yq '.operatorTests.scyllaDBVersions.upgradeFrom' "$CONFIG_FILE")}"
+
```
Comment (rzetelskik) [inline hack/.ci/lib/e2e.sh:244]: nit: move to top of the file, where `SO_BUCKET_NAME` and other default assignments are

File: hack/.ci/lib/e2e.sh [hunk 14]
```diff
@@ -234,6 +234,14 @@ function run-e2e {
   ingress_custom_annotations='haproxy.org/ssl-passthrough=true,route.openshift.io/termination=passthrough'
   ingress_controller_address="$( kubectl -n=haproxy-ingress get svc haproxy-ingress --template='{{ .spec.clusterIP }}' ):9142"
 
+  CONFIG_FILE="$(realpath "$(dirname "${BASH_SOURCE[0]}")/../../../assets/config/config.yaml")"
```
Comment (rzetelskik) [inline hack/.ci/lib/e2e.sh:237]: nit: it's a local variable so please use a lowercase name for consistency

**Review by rzetelskik — COMMENTED** [review]
thanks for the updates, a few more comments, mostly nits

File: hack/.ci/lib/e2e.sh (hunk 14 above)
Comment (grzywin) [reply hack/.ci/lib/e2e.sh]: done

File: pkg/cmd/tests/options.go [hunk 15]
```diff
@@ -164,6 +181,31 @@ func (o *TestFrameworkOptions) Validate(args []string) error {
...
+	var tagWithOptionalDigestRegexp = regexp.MustCompile("^" + referenceTagRegexp + "(?:@" + digestRegexp + ")?$")
```
[the remark this answers is not in the document]
Comment (grzywin) [reply pkg/cmd/tests/options.go]: done

File: hack/.ci/lib/e2e.sh (hunk 13 above)
Comment (grzywin) [reply hack/.ci/lib/e2e.sh]: done

File: pkg/cmd/tests/options.go (hunk 12 above)
Comment (grzywin) [reply pkg/cmd/tests/options.go]: done

File: pkg/cmd/tests/options.go [hunk 16]
```diff
@@ -164,6 +191,41 @@ func (o *TestFrameworkOptions) Validate(args []string) error {
 		errors = append(errors, fmt.Errorf("gcs-service-account-key-path and s3-credentials-file-path can't be set simultanously"))
 	}
 
+	if !tagWithOptionalDigestRegexp.MatchString(o.ScyllaDBVersion) {
+		errors = append(errors, fmt.Errorf(
+			"invalid scylladb-version format: %q. Expected format: <tag>[@<digest>], e.g., '6.2.0' or '6.2.0@sha256:abcdef...'",
```
Comment (rzetelskik) [inline pkg/cmd/tests/options.go:196]: nit: imo `<tag>[@<digest>]` is sufficient, but if you prefer to leave the example it's ok

File: test/e2e/set/scyllacluster/scyllacluster_upgrades.go [hunk 17]
```diff
@@ -5,10 +5,8 @@ package scyllacluster
 import (
 	"context"
 	"fmt"
-
```
Comment (rzetelskik) [inline test/e2e/set/scyllacluster/scyllacluster_upgrades.go:8]: shouldn't be deleted

File: pkg/cmd/tests/options.go [hunk 18]
```diff
@@ -55,8 +57,23 @@ type TestFrameworkOptions struct {
...
 
+const (
+	// https://github.com/distribution/reference/blob/8c942b0459dfdcc5b6685581dd0a5a470f615bff/regexp.go#L68
+	referenceTagRegexp = `[\w][\w.-]{0,127}`
+
+	// https://github.com/distribution/reference/blob/8c942b0459dfdcc5b6685581dd0a5a470f615bff/regexp.go#L81
+	digestRegexp = `[A-Za-z][A-Za-z0-9]*(?:[-_+.][A-Za-z][A-Za-z0-9]*)*[:][[:xdigit:]]{32,}`
+)
+
+var tagWithOptionalDigestRegexp = regexp.MustCompile("^" + referenceTagRegexp + "(?:@" + digestRegexp + ")?$")
```
Comment (rzetelskik) [inline pkg/cmd/tests/options.go:75]: nit: conventionally at the top of the file

File: pkg/cmd/tests/options.go [hunk 19]
```diff
@@ -2,8 +2,10 @@ package tests
 
 import (
 	"fmt"
+	configassets "github.com/scylladb/scylla-operator/assets/config"
```
Comment (rzetelskik) [inline pkg/cmd/tests/options.go:5]: shouldn't be grouped with std imports
